engine/execution/computation/computer/uploader/retryable_uploader.go

```diff
@@ -0,0 +1,262 @@
+package uploader
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/mempool/entity"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/storage"
+)
+
+// RetryableUploader defines the interface for uploader that is retryable
+type RetryableUploader interface {
+	Uploader
+	RetryUpload() error
+}
+
+// BadgerRetryableUploader is the BadgerDB based implementation to RetryableUploader
+type BadgerRetryableUploader struct {
+	uploader           Uploader
+	execDataService    state_synchronization.ExecutionDataService
+	unit               *engine.Unit
+	metrics            module.ExecutionMetrics
+	blocks             storage.Blocks
+	commits            storage.Commits
+	transactionResults storage.TransactionResults
+	uploadStatusStore  storage.ComputationResultUploadStatus
+}
+
+func NewBadgerRetryableUploader(
+	uploader Uploader,
+	blocks storage.Blocks,
+	commits storage.Commits,
+	transactionResults storage.TransactionResults,
+	uploadStatusStore storage.ComputationResultUploadStatus,
+	execDataService state_synchronization.ExecutionDataService,
+	metrics module.ExecutionMetrics) *BadgerRetryableUploader {
+
+	// check storage\eds interface params
+	if blocks == nil || commits == nil || transactionResults == nil ||
+		uploadStatusStore == nil || execDataService == nil {
+		panic("Not all storage parameters are valid")
+	}
+
+	// NOTE: only AsyncUploader is supported for now.
+	switch uploader.(type) {
+	case *AsyncUploader:
+		// When Uploade() is successful, the stored ComputationResult in BadgerDB will be removed.
+		onCompleteCB := func(computationResult *execution.ComputationResult, err error) {
+			if err != nil {
+				log.Warn().Msg(fmt.Sprintf("ComputationResults upload failed with ID %s",
+					computationResult.ExecutionDataID))
+				return
+			}
+
+			if computationResult == nil {
+				log.Warn().Msg("nil ComputationResult parameter")
+				return
+			}
+
+			// Removing existing upload status from BadgerDB
+			if err = uploadStatusStore.Remove(computationResult.ExecutionDataID); err != nil {
+				log.Warn().Msg(fmt.Sprintf(
+					"ComputationResults with ID %s failed to be removed on local storage. ERR: %s ",
+					computationResult.ExecutionDataID.String(), err.Error()))
+			}
+			// Store upload status as Done(true) anyway
+			if err := uploadStatusStore.Store(computationResult.ExecutionDataID,
```

Comment:
Why do we remove and insert instead of just updating?
Isn't that creating a potential race condition where uploaded entry can be tried because its not in the database?

Reply:
good point. I added an upsert implementation into storage layer so that we will not have remove-then-store issue here. However in our low-level Badger implementation, `update` code (https://github.com/onflow/flow-go/blob/master/storage/badger/operation/common.go#L88) will err out with non-existed keys, so in my storage layer code, I use both `retrieve` first to check existence of given key to implement `upsert`.

In terms of race condition, my question is how often\possible it is that the same block upload is being retried and computed (including original upload call)? if it can happen frequently, we may have to address that race-condition here.